src/transformers/generation/beam_search.py

```diff
@@ -892,6 +884,9 @@ def is_done(self, best_sum_logprobs: float, cur_len: int) -> bool:
         elif self.early_stopping:
             return True
         else:
-            cur_score = best_sum_logprobs / cur_len**self.length_penalty
+            if self.length_penalty > 0.0:
+                cur_score = best_sum_logprobs / self.max_length**self.length_penalty
+            else:
+                cur_score = best_sum_logprobs / cur_len**self.length_penalty
```

Comment:
Implements the logic as described in the PR header. `max_length` is now needed as an input.